File: bot.py
from datetime import datetime

class Config:
    def __init__(self,
        base_order_size=10.0,
        safety_order_size=20.0,
        target_profit_perc=1.0,
        price_deviation_safety_orders=0.5,
        safety_order_volume_scale=2.0,
        safety_order_step_scale=1.0) -> None:
            self.base_order_size = base_order_size
            self.safety_order_size=safety_order_size
            self.target_profit_perc=target_profit_perc
            self.price_deviation_safety_orders=price_deviation_safety_orders
            self.safety_order_volume_scale=safety_order_volume_scale
            self.safety_order_step_scale=safety_order_step_scale
            self.safety_trades_count=0
            self.closed_deals_count=0
            self.max_price_deviation=0
            self.max_price_deviation_date_max=None
            self.max_price_deviation_date_min=None
            self.max_real_safety_order_price_deviation=0.0
            self.max_local_price_deviation=0.0
            self.total=0
        
    def __str__(self) -> str:
        return ("Total: $%d\n" % self.total +
        "Base Order Size: $%.2f\n" % self.base_order_size +
        "Safety Order Size: $%.2f\n" % self.safety_order_size +
        "Target Profit Perc: %.1f%%\n" % self.target_profit_perc +
        "Price Deviation Safety Orders: %.2f\n" % self.price_deviation_safety_orders +
        "Safety Order Volume Scale: %.2f\n" % self.safety_order_volume_scale + 
        "Safety Order Step Scale: %.2f\n" % self.safety_order_step_scale +
        "Safety Trades Count: %d\n" % self.safety_trades_count +
        "Number of closed deals: %d\n" % self.closed_deals_count +
        "Max Safety Order Price Deviation: %.2f%%\n" % self.max_safety_order_price_deviation() +
        "Max Real Safety Order Price Deviation: %.2f%%\n" % self.max_real_safety_order_price_deviation +
        "Max Price Deviation: %.2f%% (%s - %s)\n" % (self.max_price_deviation, datetime.fromtimestamp(self.max_price_deviation_date_max), datetime.fromtimestamp(self.max_price_deviation_date_min)) +
        "Max Local Price Deviation: %.2f%%\n" % self.max_local_price_deviation)


    def max_safety_order_price_deviation(self):
        i = 1
        max_deviation = self.price_deviation_safety_orders
        while i < self.safety_trades_count:
            max_deviation = self.price_deviation_safety_orders + self.safety_order_step_scale * max_deviation
            i += 1

        return max_deviation

class Bot:
    def __init__(self,
        config=Config(),
        verbose=False) -> None:
            self.config = config
            self.initial_usd = 1000.0
            self.total_usd = 1000.0
            self.total_btc = 0
            self.verbose = verbose
            self.initial_buy_price = 0
            self.last_buy_price = 0

    def buy(self, amount_usd, price, date):
        amount_btc = amount_usd / price
        self.total_btc += amount_btc
        self.total_usd -= amount_usd
        if self.verbose:
            print("Buying %f bitcoins at $%.2f with $%.2f on %s" %(amount_btc, price, amount_usd, datetime.fromtimestamp(date)))
        return amount_btc   # returns the number of bitcoins that were just bought

    def sell(self, amount_btc, price, date):
        amount_usd = amount_btc * price
        self.total_usd += amount_usd
        self.total_btc -= amount_btc
        if self.verbose:
            print("Selling %f bitcoins at $%.2f for $%.2f on %s" %(amount_btc, price, amount_usd, datetime.fromtimestamp(date)))
        return amount_usd   # returns the number of dollars acquired by the sell

    def calculate_gains_usd(self):
        return self.total_usd - self.initial_usd

    def calculate_deviation(self, last_price, initial_price):
        return 100 - (last_price * 100 / initial_price)

    def simulate(self, prices, dates, start_index):
        calculate_deviation = self.calculate_deviation
        buy = self.buy
        sell = self.sell

        order_volume_usd = total_volume_usd = self.config.base_order_size
        amount_btc = buy(order_volume_usd, prices[start_index], dates[start_index])
        bought_price = prices[start_index]
        self.initial_buy_price = bought_price
        max_price = min_price = bought_price
        safety_trades_count = 0
        price_deviation_safety_orders = self.config.price_deviation_safety_orders
        all_prices = [bought_price]
        all_prices_append = all_prices.append
        temporal_index_max_deviation = start_index

        # after the first buy, the next order will be the first safety one
        order_volume_usd = self.config.safety_order_size

        # TODO: Should I convert prices to type List() ?
        for index, price in enumerate(prices):
            index += start_index
            if index == start_index:
                continue

            all_prices_append(price)

            if max_price < price:
                max_price = price
                min_price = price
                temporal_index_max_deviation = index

            if min_price > price:
                min_price = price
                deviation = calculate_deviation(min_price, max_price)
                if self.config.max_price_deviation < deviation:
                    self.config.max_price_deviation_date_max = dates[temporal_index_max_deviation]
                    self.config.max_price_deviation_date_min = dates[index]
                    self.config.max_price_deviation = deviation
            
            current_value_usd = amount_btc * price
            #print("Current value in USD=$%.2f | Price=$%.2f" %(current_value_usd, price))

            profit = current_value_usd - total_volume_usd
            expected_profit = total_volume_usd * self.config.target_profit_perc / 100

            if profit >= expected_profit:
                if self.verbose:
                    print("Total Volume USD (Total Spent)=$%.2f" %(total_volume_usd))
                amount_usd = sell(amount_btc, price, dates[index])
                self.config.safety_trades_count = max(self.config.safety_trades_count, safety_trades_count)
                safety_orders_deviation = calculate_deviation(self.last_buy_price, self.initial_buy_price)
                deviation = calculate_deviation(min(all_prices), self.initial_buy_price)
                self.config.max_local_price_deviation = max(self.config.max_local_price_deviation, deviation)
                self.config.max_real_safety_order_price_deviation = max(self.config.max_real_safety_order_price_deviation, safety_orders_deviation)
                self.config.closed_deals_count += 1
                if self.verbose:
                    print("Profit=$%f | Expected Profit=$%f" %(profit, expected_profit))
                return index, amount_usd

            if price <= bought_price  * ( 1 - price_deviation_safety_orders / 100):
                if order_volume_usd <= self.total_usd:
                    amount_btc += buy(order_volume_usd, price, dates[index])
                    bought_price = price
                    total_volume_usd += order_volume_usd
                    order_volume_usd *= self.config.safety_order_volume_scale
                    safety_trades_count += 1
                    price_deviation_safety_orders *= self.config.safety_order_step_scale
                    self.last_buy_price = price
                else:
                    continue

        if self.verbose:
            print("END of data history")
        return None, None

    def execute(self, prices, dates):
        index = -1
        simulate = self.simulate
        verbose = self.verbose
        while True:
            index, usd_amount = simulate(prices[index+1:], dates[index+1:], index+1)
            if index == None:
                break
            if verbose:
                print("Amount USD returned = $%f" % usd_amount)
                print("Total USD = $%f" % self.total_usd)
                print("Index=%d" %index)

File: go/simulator/bot.go
package simulator

import (
	"fmt"
	"math"

	"github.com/rxmxn/bitcoin_bot_simulator/go/utils"
)

type Bot struct {
	Conf            Config
	InitialUsd      float64
	TotalUsd        float64
	TotalBtc        float64
	Verbose         bool
	InitialBuyPrice float64
	LastBuyPrice    float64
}

func (bot *Bot) Buy(amount_usd float64, price float64, date int64) float64 {
	amount_btc := amount_usd / price
	bot.TotalBtc += amount_btc
	bot.TotalUsd -= amount_usd

	if bot.Verbose {
		fmt.Printf("Buying %f bitcoins at $%.2f with $%.2f on %s\n", amount_btc, price, amount_usd, utils.DateTimeToString(date))
	}

	return amount_btc
}

func (bot *Bot) Sell(amount_btc float64, price float64, date int64) float64 {
	amount_usd := amount_btc * price
	bot.TotalUsd += amount_usd
	bot.TotalBtc -= amount_btc

	if bot.Verbose {
		fmt.Printf("Selling %f bitcoins at $%.2f with $%.2f on %s\n", amount_btc, price, amount_usd, utils.DateTimeToString(date))
	}

	return amount_usd
}

func (bot *Bot) CalculateGainsUsd() float64 {
	return bot.TotalUsd - bot.InitialUsd
}

func (bot *Bot) CalculateDeviation(last_price float64, initial_price float64) float64 {
	return 100 - (last_price * 100 / initial_price)
}

func (bot *Bot) Simulate(prices []float64, dates []int64) (int, float64) {
	order_volume_usd := bot.Conf.BaseOrderSize
	total_volume_usd := bot.Conf.BaseOrderSize

	amount_btc := bot.Buy(order_volume_usd, prices[0], dates[0])
	bought_price := prices[0]
	bot.InitialBuyPrice = bought_price
	max_price := bought_price
	min_price := bought_price
	safety_trades_count := 0

	price_deviation_safety_orders := bot.Conf.PriceDeviationSafetyOrders
	all_prices := []float64{bought_price}
	temporal_index_max_deviation := 0

	order_volume_usd = bot.Conf.SafetyOrderSize

	for index, price := range prices {
		if index == 0 {
			continue
		}

		all_prices = append(all_prices, price)

		if max_price < price {
			max_price = price
			min_price = price
			temporal_index_max_deviation = index
		}

		if min_price > price {
			min_price = price
			deviation := bot.CalculateDeviation(min_price, max_price)
			if bot.Conf.MaxPriceDeviation < deviation {
				bot.Conf.MaxPriceDeviationDateMax = dates[temporal_index_max_deviation]
				bot.Conf.MaxPriceDeviationDateMin = dates[index]
				bot.Conf.MaxPriceDeviation = deviation
			}
		}

		current_value_usd := amount_btc * price

		profit := current_value_usd - total_volume_usd
		expected_profit := total_volume_usd * bot.Conf.TargetProfitPerc / 100

		if profit >= expected_profit {
			if bot.Verbose {
				fmt.Printf("Total Volume USD (Total Spent)=$%.2f\n", total_volume_usd)
			}
			amount_usd := bot.Sell(amount_btc, price, dates[index])
			bot.Conf.SafetyTradesCount = int(math.Max(float64(bot.Conf.SafetyTradesCount), float64(safety_trades_count)))
			safety_orders_deviation := bot.CalculateDeviation(bot.LastBuyPrice, bot.InitialBuyPrice)
			min_all_prices, _ := utils.MinMax(all_prices)
			deviation := bot.CalculateDeviation(min_all_prices, bot.InitialBuyPrice)
			bot.Conf.MaxLocalPriceDeviation = math.Max(bot.Conf.MaxLocalPriceDeviation, deviation)
			bot.Conf.MaxRealSafetyOrderPriceDeviation = math.Max(bot.Conf.MaxRealSafetyOrderPriceDeviation, safety_orders_deviation)
			bot.Conf.ClosedDealsCount += 1
			if bot.Verbose {
				fmt.Printf("Profit=$%f | Expected Profit=$%f\n", profit, expected_profit)
			}
			return index, amount_usd
		}

		if price <= bought_price*(1-price_deviation_safety_orders/100) {
			if order_volume_usd <= bot.TotalUsd {
				amount_btc += bot.Buy(order_volume_usd, price, dates[index])
				bought_price = price
				total_volume_usd += order_volume_usd
				order_volume_usd *= bot.Conf.SafetyOrderVolumeScale
				safety_trades_count += 1
				price_deviation_safety_orders *= bot.Conf.SafetyOrderStepScale
				bot.LastBuyPrice = price
			} else {
				continue
			}
		}
	}

	if bot.Verbose {
		fmt.Println("END of data history")
	}

	return -1, -1
}

func (bot *Bot) Execute(prices []float64, dates []int64) {
	last_price := prices[len(prices)-1]

	for {
		index, usd_amount := bot.Simulate(prices, dates)
		if index == -1 {
			break
		}
		dates = dates[index+1:]
		prices = prices[index+1:]
		if bot.Verbose {
			fmt.Printf("Amount USD returned = $%f\n", usd_amount)
			fmt.Printf("Total USD = $%f\n", bot.TotalUsd)
			fmt.Printf("Index=%d\n", index)
		}
	}

	bot.Conf.Total = bot.TotalUsd + bot.TotalBtc*last_price
}
